File: src/visual-graph/builder/find-host-subgraph.test.ts
import { describe, expect, test } from "vitest";

import { asScopeId } from "../../ir/serialized/scope-id.js";
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import { DIRECTION } from "../direction.js";
import { SUBGRAPH_KIND } from "../subgraph-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualSubgraph } from "../visual-subgraph.js";
import type { BuildState } from "./build-state.js";
import { findHostSubgraph } from "./find-host-subgraph.js";
import { baseRef } from "./testing/make-ref.js";
import { baseScope } from "./testing/make-scope.js";

function baseSubgraph(id: string): VisualSubgraph {
  return {
    type: VISUAL_ELEMENT_TYPE.Subgraph,
    id,
    kind: SUBGRAPH_KIND.Function,
    line: 1,
    endLine: null,
    direction: DIRECTION.RL,
    ownerNodeId: "n_owner",
    ownerName: "owner",
    elements: [],
  };
}

function emptyState(overrides: Partial<BuildState> = {}): BuildState {
  return {
    subgraphByScope: new Map(),
    functionSubgraphByFn: new Map(),
    returnSubgraphsByFn: new Map(),
    returnUseAdded: new Set(),
    ifTestAnchorByOffset: new Map(),
    switchDiscriminantAnchorByOffset: new Map(),
    whileTestAnchorByOffset: new Map(),
    doWhileTestAnchorByOffset: new Map(),
    forTestAnchorByOffset: new Map(),
    pendingLoopTestAnchors: [],
    expressionStatementByOffset: new Map(),
    emittedEdges: new Set(),
    edges: [],
    ...overrides,
  };
}

const root = { ...baseScope(), id: asScopeId("root") };
const inner = {
  ...baseScope(),
  id: asScopeId("inner"),
  upper: asScopeId("root"),
};
const leaf = {
  ...baseScope(),
  id: asScopeId("leaf"),
  upper: asScopeId("inner"),
};
const scopeMap = new Map<string, SerializedScope>(
  [root, inner, leaf].map((v) => [v.id, v]),
);

describe("findHostSubgraph", () => {
  test("returns the subgraph mapped to the ref's own scope when present", () => {
    const sg = baseSubgraph("s_leaf");
    const state = emptyState({ subgraphByScope: new Map([["leaf", sg]]) });
    const ref = { ...baseRef(), from: asScopeId("leaf") };
    expect(findHostSubgraph(ref, "fnVar", scopeMap, state)).toEqual(sg);
  });

  test("walks up via .upper to find the closest enclosing subgraph", () => {
    const sg = baseSubgraph("s_root");
    const state = emptyState({ subgraphByScope: new Map([["root", sg]]) });
    const ref = { ...baseRef(), from: asScopeId("leaf") };
    expect(findHostSubgraph(ref, "fnVar", scopeMap, state)).toEqual(sg);
  });

  test("falls back to the function subgraph for the enclosing fn variable id", () => {
    const fnSg = baseSubgraph("s_fn");
    const state = emptyState({
      functionSubgraphByFn: new Map([["fnVar", fnSg]]),
    });
    const ref = { ...baseRef(), from: asScopeId("leaf") };
    expect(findHostSubgraph(ref, "fnVar", scopeMap, state)).toEqual(fnSg);
  });

  test("returns null when neither chain nor fn fallback yields a subgraph", () => {
    const ref = { ...baseRef(), from: asScopeId("leaf") };
    expect(findHostSubgraph(ref, "nope", scopeMap, emptyState())).toEqual(null);
  });

  test("returns null when ref.from is not in the scope map and the fn fallback is also missing", () => {
    const ref = { ...baseRef(), from: asScopeId("missing") };
    expect(findHostSubgraph(ref, "missingFn", scopeMap, emptyState())).toEqual(
      null,
    );
  });

  test("returns the scope-chain subgraph even when enclosingFnVarId is null", () => {
    const sg = baseSubgraph("s_root");
    const state = emptyState({ subgraphByScope: new Map([["root", sg]]) });
    const ref = { ...baseRef(), from: asScopeId("leaf") };
    expect(findHostSubgraph(ref, null, scopeMap, state)).toEqual(sg);
  });

  test("returns null when enclosingFnVarId is null and the scope chain yields no subgraph", () => {
    const ref = { ...baseRef(), from: asScopeId("leaf") };
    const fnSg = baseSubgraph("s_fn");
    const state = emptyState({
      // A function subgraph exists but it must be ignored when fnVarId is null.
      functionSubgraphByFn: new Map([["fnVar", fnSg]]),
    });
    expect(findHostSubgraph(ref, null, scopeMap, state)).toEqual(null);
  });
});

File: crates/unsnarl-visual-graph/src/builder/find_host_subgraph_test.rs
//! Sibling tests for [`find_host_subgraph`].
//! The
//! Rust signature returns a [`SubgraphIdx`] handle (vs the TS
//! `VisualSubgraph` value) so each test asserts on the returned
//! handle's identity against the arena slots it populated.

use std::collections::HashMap;

use unsnarl_ir::serialized::SerializedScope;

use crate::builder::arena::{BuildArena, SubgraphIdx};
use crate::builder::find_host_subgraph::find_host_subgraph;
use crate::builder::state::BuildState;
use crate::builder::testing::{base_serialized_reference, base_serialized_scope, scope_id};
use crate::direction::Direction;
use crate::visual_element_type::SubgraphTypeTag;
use crate::visual_subgraph::{OwnedExtras, OwnedSubgraphKind, OwnedVisualSubgraph, VisualSubgraph};

fn push_function_subgraph(arena: &mut BuildArena, id: &str) -> SubgraphIdx {
    arena.push_subgraph(VisualSubgraph::Owned(OwnedVisualSubgraph {
        r#type: SubgraphTypeTag::Subgraph,
        id: id.to_string(),
        kind: OwnedSubgraphKind::Function,
        line: 1,
        end_line: None,
        direction: Direction::RL,
        extras: OwnedExtras::Function {
            owner_node_id: Some("n_owner".to_string()),
            owner_name: "owner".to_string(),
        },
        elements: Vec::new(),
    }))
}

fn build_scope_chain() -> [SerializedScope; 3] {
    let root = base_serialized_scope("root");
    let mut inner = base_serialized_scope("inner");
    inner.upper = Some(scope_id("root"));
    let mut leaf = base_serialized_scope("leaf");
    leaf.upper = Some(scope_id("inner"));
    [root, inner, leaf]
}

fn scope_map(scopes: &[SerializedScope]) -> HashMap<&str, &SerializedScope> {
    scopes.iter().map(|s| (s.id.value(), s)).collect()
}

#[test]
fn returns_subgraph_mapped_to_refs_own_scope_when_present() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let mut arena = BuildArena::new();
    let sg = push_function_subgraph(&mut arena, "s_leaf");
    let mut state = BuildState::new();
    state.subgraph_by_scope.insert("leaf".to_string(), sg);

    let mut r = base_serialized_reference();
    r.from = scope_id("leaf");

    assert_eq!(
        find_host_subgraph(&r, Some("fnVar"), &map, &state),
        Some(sg)
    );
}

#[test]
fn walks_up_via_upper_to_find_closest_enclosing_subgraph() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let mut arena = BuildArena::new();
    let sg = push_function_subgraph(&mut arena, "s_root");
    let mut state = BuildState::new();
    state.subgraph_by_scope.insert("root".to_string(), sg);

    let mut r = base_serialized_reference();
    r.from = scope_id("leaf");

    assert_eq!(
        find_host_subgraph(&r, Some("fnVar"), &map, &state),
        Some(sg)
    );
}

#[test]
fn falls_back_to_function_subgraph_for_enclosing_fn_variable_id() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let mut arena = BuildArena::new();
    let fn_sg = push_function_subgraph(&mut arena, "s_fn");
    let mut state = BuildState::new();
    state
        .function_subgraph_by_fn
        .insert("fnVar".to_string(), fn_sg);

    let mut r = base_serialized_reference();
    r.from = scope_id("leaf");

    assert_eq!(
        find_host_subgraph(&r, Some("fnVar"), &map, &state),
        Some(fn_sg)
    );
}

#[test]
fn returns_none_when_neither_chain_nor_fn_fallback_yields_subgraph() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let state = BuildState::new();

    let mut r = base_serialized_reference();
    r.from = scope_id("leaf");

    assert_eq!(find_host_subgraph(&r, Some("nope"), &map, &state), None);
}

#[test]
fn returns_none_when_ref_from_is_not_in_scope_map_and_fn_fallback_missing() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let state = BuildState::new();

    let mut r = base_serialized_reference();
    r.from = scope_id("missing");

    assert_eq!(
        find_host_subgraph(&r, Some("missingFn"), &map, &state),
        None
    );
}

#[test]
fn returns_scope_chain_subgraph_even_when_enclosing_fn_var_id_is_none() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let mut arena = BuildArena::new();
    let sg = push_function_subgraph(&mut arena, "s_root");
    let mut state = BuildState::new();
    state.subgraph_by_scope.insert("root".to_string(), sg);

    let mut r = base_serialized_reference();
    r.from = scope_id("leaf");

    assert_eq!(find_host_subgraph(&r, None, &map, &state), Some(sg));
}

#[test]
fn returns_none_when_enclosing_fn_var_id_is_none_and_scope_chain_empty() {
    let scopes = build_scope_chain();
    let map = scope_map(&scopes);
    let mut arena = BuildArena::new();
    let fn_sg = push_function_subgraph(&mut arena, "s_fn");
    let mut state = BuildState::new();
    state
        .function_subgraph_by_fn
        .insert("fnVar".to_string(), fn_sg);

    let mut r = base_serialized_reference();
    r.from = scope_id("leaf");

    assert_eq!(find_host_subgraph(&r, None, &map, &state), None);
}
